### PR TITLE
add support for adding/removing virtio devices during runtime

### PR DESCRIPTION
**Pull request by M1cha**
This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/241)

### CODE REVIEW

File: tools/lkl/lib/fs.c [hunk 1]
```diff
@@ -34,19 +37,90 @@ static uint32_t new_encode_dev(unsigned int major, unsigned int minor)
...
+	if (ret < 0 || (size_t) ret >= sizeof(sysfs_path) - sysfs_path_len)
+		return -LKL_ENOMEM;
+	sysfs_path_len += ret;
+
+	disk_name = get_node_with_prefix(sysfs_path, "vd", &opendir_ret);
+	if (!disk_name)
+		return (long)opendir_ret;
+
+	ret =
+	    snprintf(sysfs_path + sysfs_path_len,
```
Comment (tavip) [inline tools/lkl/lib/fs.c]: Here as wel, we can merge ret and sprintf on the same line

**Comment by M1cha** [comment]
@tavip I updated the commit according to your comments.

File: tools/lkl/lib/fs.c [hunk 2]
```diff
@@ -34,19 +37,81 @@ static uint32_t new_encode_dev(unsigned int major, unsigned int minor)
...
+	if (disk_id < 0)
+		return -LKL_EINVAL;
 
 	ret = lkl_mount_fs("sysfs");
 	if (ret < 0)
 		return ret;
 
-	sysfs_path[strlen("/sysfs/block/vd")] += disk_id;
+	if ((uint32_t) disk_id >= virtio_get_num_bootdevs())
+		ret = snprintf(sysfs_path, sizeof(sysfs_path), "/sysfs/devices/platform/virtio-mmio.%d.auto", disk_id - virtio_get_num_bootdevs());
```
Comment (tavip) [inline tools/lkl/lib/fs.c]: Since this line is very long (>100), you can split this one after the string, like:

ret = snprintf(sysfs_path, sizeof(sysfs_path), "/sysfs/devices/platform/virtio-mmio.%d.auto", 
                     disk_id - virtio_get_num_bootdevs());

File: tools/lkl/lib/fs.c [hunk 3]
```diff
@@ -34,19 +37,81 @@ static uint32_t new_encode_dev(unsigned int major, unsigned int minor)
...
+	else
+		ret = snprintf(sysfs_path, sizeof(sysfs_path), "/sysfs/devices/virtio-mmio-cmdline/virtio-mmio.%d", disk_id);
```
Comment (tavip) [inline tools/lkl/lib/fs.c]: Same here, you can split after the string.

File: tools/lkl/lib/fs.c (hunk 1 above)
Comment (tavip) [inline tools/lkl/lib/fs.c]: You missed this one :)

File: tools/lkl/lib/virtio.c [hunk 4]
```diff
@@ -544,24 +546,82 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
...
-	num_bytes = snprintf(devs, avail, " virtio_mmio.device=%d@0x%lx:%d",
-			 mmio_size, (uintptr_t)dev->base, dev->irq);
-	if (num_bytes < 0 || num_bytes >= avail) {
-		lkl_put_irq(dev->irq, "virtio");
-		unregister_iomem(dev->base);
-		lkl_host_ops.mem_free(dev->queue);
-		return -LKL_ENOMEM;
+	if (!lkl_is_running()) {
+		avail = sizeof(lkl_virtio_devs) - (devs - lkl_virtio_devs);
+		num_bytes =
```
Comment (tavip) [inline tools/lkl/lib/virtio.c]: Here as well

File: tools/lkl/lib/virtio.c [hunk 5]
```diff
@@ -544,24 +546,82 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
...
+
+	if (!lkl_is_running())
+		goto skip_unbind;
+
+	ret = lkl_mount_fs("sysfs");
+	if (ret < 0)
+		return ret;
+
+	if (dev->virtio_mmio_id >= virtio_get_num_bootdevs())
+		ret =
```
Comment (tavip) [inline tools/lkl/lib/virtio.c]: And here

File: tools/lkl/lib/virtio.c [hunk 6]
```diff
@@ -544,24 +546,82 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
...
+		    snprintf(devname, sizeof(devname), "virtio-mmio.%d.auto",
+			     dev->virtio_mmio_id - virtio_get_num_bootdevs());
+	else
+		ret =
```
Comment (tavip) [inline tools/lkl/lib/virtio.c]: Here as well

**Comment by M1cha** [comment]
@tavip no problem, upstreaming LKL would be really great :)

**Comment by M1cha** [comment]
btw 'drivers/virtio/virtio_mmio.c' throws warnings when removing devices about a missing remove-functions. this definitely looks like a upstream bug. I locally added a empty function to 'vm_dev->vdev.dev.release' to quiet the warnings. I dunno if and what kind of cleanup this driver needs.

File: tools/lkl/lib/virtio.c [hunk 7]
```diff
@@ -544,24 +546,79 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
...
+
+	ret = lkl_sys_write(fd, devname, strlen(devname));
+	if (ret < 0)
+		return ret;
+
+	ret = lkl_sys_close(fd);
+	if (ret < 0)
+		return ret;
+
+	ret = lkl_sys_umount("/sysfs", 0);
```
Comment (liuyuan10) [inline tools/lkl/lib/virtio.c]: If sysfs is already mounted before this function, it will umount it. Can you umount only when lkl_mount_fs returns 0?

File: tools/lkl/lib/fs.c [hunk 8]
```diff
@@ -29,19 +32,88 @@ int lkl_mount_fs(char *fstype)
...
+	struct lkl_linux_dirent64 *dirent;
+	char *result = NULL;
+
+	dir = lkl_opendir(path, ret);
+	if (!dir)
+		return NULL;
+
+	while ((dirent = lkl_readdir(dir))) {
+		if (startswith(dirent->d_name, prefix))
+			result = strdup(dirent->d_name);
```
Comment (liuyuan10) [inline tools/lkl/lib/fs.c]: break early?

File: tools/lkl/lib/virtio.c [hunk 9]
```diff
@@ -544,24 +546,79 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
...
+		if (num_bytes < 0 || num_bytes >= avail) {
+			lkl_put_irq(dev->irq, "virtio");
+			unregister_iomem(dev->base);
+			lkl_host_ops.mem_free(dev->queue);
+			return -LKL_ENOMEM;
+		}
+		devs += num_bytes;
+		dev->virtio_mmio_id = lkl_num_virtio_boot_devs++;
+	} else {
+		ret =
```
Comment (liuyuan10) [inline tools/lkl/lib/virtio.c]: There are several "Must be called before calling lkl_start_kernel" in lkl.h. Can you correct them?

Can I add virtio devs after lkl is running concurrently in multiple threads? At least seems register_iomem() is not thread safe. If not, can you make a comment somewhere to make it clear?

**Comment by thehajime** [comment]
it is more like a general question but is it possible to adapt this patch to a netdev (dynamically added new virtio-net dev by a program) ?  we may need to update tools/lkl/lib/net.c or else though.

File: tools/lkl/lib/virtio.c (hunk 9 above)
Comment (M1cha) [inline tools/lkl/lib/virtio.c]: I think it's not even threads safe before starting LKL. if that's right I think the comments should be updated in a different commit.

Comment (liuyuan10) [inline tools/lkl/lib/virtio.c]: Initially I thought this PR applies to all virtio devs but seems only disk is tested. It's fine then.
